FontSizePicker: Use components instead of helper functions

noisysocks opened the pull request:
## What?
Simplify how `FontSizePicker` is implemented by using nested components for the `CustomSelectControl` and `ToggleGroupControl` logic instead of helper functions.

## Why?
While adding types to `FontSizePicker` in https://github.com/WordPress/gutenberg/pull/44449 I ran into difficulty typing `getFontSizeOptions` as its return type depends on the runtime value of `useSelectControl`. See https://github.com/WordPress/gutenberg/pull/44449#discussion_r980489635 for some relevant discussion there.

The end result is lots of yucky `as` casts scattered throughout the component:

https://github.com/WordPress/gutenberg/blob/3983fee52b8fa0b9336d6202e8b92d407b99187c/packages/components/src/font-size-picker/index.tsx#L218

In addition to being difficult to type, it's just plain confusing. @ramonjd learned this the other day when trying to make a simple change to `FontSizePicker` in https://github.com/WordPress/gutenberg/pull/44791. He only needed to change one line but the design of this component tricked him into thinking it was more complicated than it was. (Sorry to pick on you @ramonjd.)

To do away with all of this, I've settled on  _embracing_ that `FontSizePicker` works differently depending on if there are more than 5 font sizes or not. I've split out the `CustomSelectControl` (>5 font sizes) and `ToggleGroupControl` (≤5 font sizes) into their own components. Now each component can do whatever it needs to do without having to worry about the other. We can then remove the difficult-to-type helper functions altogether.

## How?
- New `FontSizePickerSelect` component: this contains the `CustomSelectControl` and logic for when there are > 5 font sizes.
- New `FontSizePickerToggleGroup` component: this contains the `ToggleGroupControl` and logic for when there are ≤5 font sizes.
- Removed no-longer-necessary helpers from `utils.ts`.

## Testing Instructions
Either:

- Look at `FontSizePicker` in Storybook, or;
- Go to Site Editor → Global Styles → Typography → Text → Size.

The font size picker should work exactly as it does in `trunk` and tests should pass.

noisysocks committed:
FontSizePicker: Use components instead of helper functions

Simplify how FontSizePicker is implemented by using nested components
instead of helper functions. Components are easier to reason about and
much easier to type.

noisysocks committed:
Remove duplicate dependency header

noisysocks commented on packages/components/src/font-size-picker/test/utils.ts:
```diff
@@ -80,74 +76,3 @@ describe( 'splitValueAndUnitFromSize', () => {
 	);
 } );
-
-describe( 'getToggleGroupOptions', () => {
```
I don't think these tests add much value beyond what's already tested in `test/index.tsx`.

ciampo replied:
I had a look and couldn't really see a test checking for t-shirt sizes. Could you add a test for that in `test/index.tsx` ?

noisysocks replied:
👍 added a test that checks all the t-shirt labels are used even when there's no font size names.

ciampo commented on packages/components/src/font-size-picker/types.ts:
```diff
@@ -81,18 +76,25 @@ export type FontSize = {
+	size: FontSizePickerProps[ 'size' ];
+	onChange: ( value: number | string | undefined ) => void;
+	onSelectCustom: () => void;
 };
```
We can reuse more prop types from `FontSizePickerProps` by using the `NonNullable` utility, and we can use `Pick` for more clarity:

```suggestion
export type FontSizePickerSelectProps = Pick<
	FontSizePickerProps,
	'value' | 'size'
> & {
	fontSizes: NonNullable< FontSizePickerProps[ 'fontSizes' ] >;
	onChange: NonNullable< FontSizePickerProps[ 'onChange' ] >;
	onSelectCustom: () => void;
};
```

ciampo commented on packages/components/src/font-size-picker/types.ts:
```diff
@@ -81,18 +76,25 @@ export type FontSize = {
+	value: FontSizePickerProps[ 'value' ];
+	__nextHasNoMarginBottom?: boolean;
+	size: FontSizePickerProps[ 'size' ];
+	onChange: ( value: number | string | undefined ) => void;
```
Same as previous comment

```suggestion
export type FontSizePickerToggleGroupProps = Pick<
	FontSizePickerProps,
	'value' | 'size' | '__nextHasNoMarginBottom'
> & {
	fontSizes: NonNullable< FontSizePickerProps[ 'fontSizes' ] >;
	onChange: NonNullable< FontSizePickerProps[ 'onChange' ] >;
```

ciampo commented on packages/components/src/font-size-picker/select.tsx:
```diff
@@ -0,0 +1,74 @@
+	const options: FontSizePickerSelectOption[] = [
+		defaultOption,
+		...fontSizes.map( ( fontSize ) => {
+			const [ number ] = splitValueAndUnitFromSize( fontSize.size );
+			return {
+				key: fontSize.slug,
+				name: fontSize.name ?? fontSize.slug,
+				value: fontSize.size,
+				__experimentalHint: number,
```
Minor nit: let's not call this variable `number`, since its type is actually a `string` (my brain refused to process this for a while 😅 )

```suggestion
			const [ parsedValue ] = splitValueAndUnitFromSize( fontSize.size );
			return {
				key: fontSize.slug,
				name: fontSize.name ?? fontSize.slug,
				value: fontSize.size,
				__experimentalHint: parsedValue,
```

noisysocks replied:
In https://github.com/WordPress/gutenberg/pull/44598 I changed this function to actually return a number. But yeah for now makes sense to name it what it is.

ciampo commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -94,28 +82,16 @@ const UnforwardedFontSizePicker = (
-				disableCustomFontSizes
-			),
-		[ shouldUseSelectControl, fontSizes, disableCustomFontSizes ]
+	const selectedFontSize = fontSizes.find(
+		( fontSize ) => fontSize.size === value
```
Previously, the `getSelectedOption` utility function used to fallback to `DEFAULT_FONT_SIZE_OPTION` when `value` was not defined, and to fallback to `CUSTOM_FONT_SIZE_OPTION` when there were not `fontSizes` matching the `value`.

How is that piece of logic implemented in this new version of the component?

noisysocks replied:
That logic only ever had any effect when `shouldUseSelectControl` was `true` and the `CustomSelectControl` was rendered, because `ToggleGroupControl` used `value={ value }` and not `value={ selectedOption.size }`. I therefore moved this logic to `FontSizePickerSelect` so that it is near to where it's needed and not executed unnecessarily. 

https://github.com/WordPress/gutenberg/blob/bbf17a3f5c90f488bbb9b6b34146d5b3cadca859/packages/components/src/font-size-picker/select.tsx#L42-L43

ciampo replied:
Thank you for the explanation! This makes sense for the fallback to `customOption`, but I'm still confused about where this piece of logic is implemented in this new version: 

https://github.com/WordPress/gutenberg/blob/d20cc8ba7e84d2385b1e257ac8ef397a515a66a8/packages/components/src/font-size-picker/utils.ts#L147-L149

ciampo replied:
Maybe I get it — when `value` is `undefined`, this code will actually select the default option (since it also have a value of `undefined`):


https://github.com/WordPress/gutenberg/blob/bbf17a3f5c90f488bbb9b6b34146d5b3cadca859/packages/components/src/font-size-picker/select.tsx#L42-L43

If that's the reason, I must admit that the logic here is much less explicit than it used to be, and I would argue that it would greatly benefit from a unit test to make sure that we don't introduce a regression in the future without realizing.

ciampo replied:
I also realised that when the `value` prop is `undefined` the header hint:
 - is `"Default"` on `trunk`
 - but it's an empty string in this PR

noisysocks replied:
> Maybe I get it — when `value` is `undefined`, this code will actually select the default option (since it also have a value of `undefined`):

Yes that's right. I'll make the `find()` code more explicit with a ternary.

> I also realised that when the `value` prop is `undefined` the header hint:

👍 I've fixed this and restored most of the changes I made to `headerHint`. Got a bit ahead of myself there, I'll wait for https://github.com/WordPress/gutenberg/pull/44966 which has more substantial changes to the hint text.

ciampo replied:
> Yes that's right. I'll make the find() code more explicit with a ternary.

Looking more into this, I found a couple more more differences with `trunk`

1. With more than 5 font sizes and when passing `value=""` to `FontSizePicker`, prior to your latest change that introduces the ternary, the `selectedOption` would have been `CUSTOM_OPTION` instead of `DEFAULT_OPTION` — that's because the empty string wouldn't match against the `undefined` value of the default option. This has been fixed by using the `value ?` ternary, but I though I'd flag it nonetheless
2. One difference that still remains between this PR and `trunk` is that, when `value` is `""`:
   - on `trunk`, the component shows a header hint that says "size default" and:
     - with 5+ options, the `CustomSelectControl` with the default option selected (and 
     - with less than 5 options, the `ToggleGroupControl` without no options selected 
   - in this PR, the component shows an empty custom value control input and the header hint says "size (custom)"

| Story (with `value=""`) | `trunk` | This PR |
|---|---|---|
| Default | ![Screenshot 2022-11-01 at 13 43 10](https://user-images.githubusercontent.com/1083581/199235356-c6d6ef6b-64e5-4f8f-9947-aa126c60d05e.png) | ![Screenshot 2022-11-01 at 13 43 56](https://user-images.githubusercontent.com/1083581/199235389-1753f343-7feb-4f24-ab2f-265e345a60d3.png) |
| With More Font Sizes | ![Screenshot 2022-11-01 at 13 43 19](https://user-images.githubusercontent.com/1083581/199235467-d8647de9-1b1d-4e40-a8f8-97ae5e467675.png) | ![Screenshot 2022-11-01 at 13 43 56](https://user-images.githubusercontent.com/1083581/199235490-bd05796d-87fa-4d47-bd9a-8cd093255ef5.png) |

After fixing those discrepancies, I also suggest that we add unit tests to protect against these regressions in the future!

ciampo replied:
Fixed in https://github.com/WordPress/gutenberg/pull/44891/commits/639db98a1e4bacd081de03609b57524f9fbb8507

ciampo commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -124,51 +100,49 @@ const UnforwardedFontSizePicker = (
-			);
+			return value !== undefined && isSimpleCssValue( value )
+				? `(${ value })`
+				: '';
```
Super nitty: probably `undefined` or `null` are a better value for when there's no `headerHint` to render than an empty string?

But really, this is mostly down to your preference — changing from `false` to `''` is already a nice improvement

noisysocks replied:
I like that this way the type is `headerHint: string`. Good old string. Does what it says on the tin.

ciampo commented on packages/components/src/font-size-picker/select.tsx:
```diff
@@ -0,0 +1,74 @@
+	];
+
+	const selectedOption =
+		options.find( ( option ) => option.value === value ) ?? customOption;
```
Previously, the `selectedOption` was computed by matching `key` and `slug`:

```js
option.key === selectedOption.slug
```

By matching the option's value instead, there's a risk of matching the wrong options, in cese two different options share the same value. I believe we should revert to using `slug`s instead

noisysocks replied:
Yes it's true that we grabbed the option with `option.key === selectedOption.slug` to pass to `value`:

https://github.com/WordPress/gutenberg/blob/d7e9b2904b188ddbe8a7edc753739cca5438b834/packages/components/src/font-size-picker/index.tsx#L219-L224

But drill down into where `selectedOption` came from and you'll see that it's the font size where `fontSize.size === value`:

https://github.com/WordPress/gutenberg/blob/d7e9b2904b188ddbe8a7edc753739cca5438b834/packages/components/src/font-size-picker/utils.ts#L151

This layer of indirection is now gone.

ciampo replied:
I see, thank you for the explanation!

I still wonder if this `value`-base check is too brittle and we're going to end up in a situation [like the `ColorPalette` one](https://github.com/WordPress/gutenberg/issues/43197) — but this definitely seems out of scope for this PR.

ciampo commented on packages/components/src/font-size-picker/select.tsx:
```diff
@@ -0,0 +1,74 @@
+	const customOption: FontSizePickerSelectOption = {
+		key: 'custom',
+		name: __( 'Custom' ),
+	};
```
Nit: These values are constants and should be defined outside of the `FontSizePickerSelect` function

ciampo commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -209,64 +183,32 @@ const UnforwardedFontSizePicker = (
 								} }
-								size={ size }
+								onSelectCustom={ () =>
+									setShowCustomValueControl( true )
```
Previously, the `onChange` callback prop was getting called also when the `custom` select option was chosen. With the new changes, it looks like `onChange` won't be fired.

Although previously, when fired, the value passed as an argument to `onChange` seemed to be `NaN` — can you confirm too? In that case, we could treat this as a bug fix.

noisysocks replied:
The NaN bug I actually fixed in https://github.com/WordPress/gutenberg/pull/44598/files#r986406892.

You're right that this no longer calls `onChange( undefined )` when switching to Custom. I think that this is correct though. We don't call `onChange( undefined )` when using the toggle button in the top right, so why should we when using the dropdown?

ciampo replied:
Sounds good to me. Afterall, if the actual `value` doesn't change, I don't see the point of calling `onChange`

ciampo commented on packages/components/src/font-size-picker/select.tsx:
```diff
@@ -0,0 +1,74 @@
+				__experimentalHint: number,
+			};
+		} ),
+		customOption,
```
Previously, `customOption` was being added only when `disableCustomFontSizes === false` (see the deleted `getSelectOptions` function). We should probably add that check back

noisysocks replied:
Oops, good catch!

ciampo commented on packages/components/src/font-size-picker/types.ts:
```diff
@@ -81,18 +76,25 @@ export type FontSize = {
-	value: number | string;
-	label: string;
 	name: string;
+	value?: FontSize[ 'size' ];
```
Is there any particular reason why this prop name was changed from `size` to `value` ? It doesn't look like this affects `CustomSelectControl`, but was curious to know the reason.

noisysocks replied:
Two reasons.

1) These things are called `value` once given to `onChange`:


     https://github.com/WordPress/gutenberg/blob/bbf17a3f5c90f488bbb9b6b34146d5b3cadca859/packages/components/src/font-size-picker/types.ts#L25

2) It makes it consistent with how the `ToggleGroupControl` works where each `ToggleGroupControlOption` has a `value` prop:

   https://github.com/WordPress/gutenberg/blob/bbf17a3f5c90f488bbb9b6b34146d5b3cadca859/packages/components/src/font-size-picker/toggle-group.tsx#L48

Hopefully soon we can make `CustomSelectControl` work more like `ToggleGroupControl` in that we have a `CustomSelectControlOption` component. It would make `FontSizePickerSelect` much more comprehensible imo 😀

ciampo replied:
> Hopefully soon we can make CustomSelectControl work more like ToggleGroupControl in that we have a CustomSelectControlOption component. It would make FontSizePickerSelect much more comprehensible imo 😀

I hope I'll be able to resume the work on `CustomSelectControl` soon enough, although I'm not sure if we're going to be able to have `option`-like children, given the data-processing (ie. filtering) needs of the component (see [this comment](https://github.com/WordPress/gutenberg/pull/41952#issuecomment-1171140086))

ciampo commented on packages/components/src/font-size-picker/toggle-group.tsx:
```diff
@@ -0,0 +1,58 @@
+					key={ fontSize.slug }
+					value={ fontSize.size }
+					label={ FONT_SIZES_ALIASES[ index ] }
+					aria-label={ fontSize.name }
```
We should preserve the same logic implemented in the old `getToggleGroupOptions`

```suggestion
					aria-label={ fontSize.name || FONT_SIZES_ALIASES[ index ] }
```

noisysocks replied:
Oops yep looks like I mixed up `label` and `aria-label`.

ciampo reviewed (commented):
Thank you for working on this, @noisysocks !

This has been a much deeper review than anticipated 😅 

Apart from the inline comments, could we also rename:

 - `toggle-group.tsx` to `font-size-picker-toggle-group.tsx`
 - `select.tsx` to `font-size-picker-select.tsx`


The following observations are also present on `trunk`, but I thought I'd mention them anyway:

1. I noticed that the user can still enter a custom value even if `disableCustomFontSizes` is set to `true`, when `withSlider` is also set to `true`
2. The logic around `fallbackFontSize`, `withReset` and `withSlider` is very confusing and should be reviewed.
3. The component can get into a weird state if:
   i. Click on the toggle to show the custom font size controls
   ii. Change the `disableCustomFontSizes` to `true`
   iii. The UI disappears and the user is stuck

noisysocks commented:
Thanks for the feedback @ciampo!

> 1. I noticed that the user can still enter a custom value even if `disableCustomFontSizes` is set to `true`, when `withSlider` is also set to `true`

I think https://github.com/WordPress/gutenberg/pull/44598 should accidentally fix this!

> 2. The logic around `fallbackFontSize`, `withReset` and `withSlider` is very confusing and should be reviewed.

The whole component is very confusing 😂. I'd be keen to deprecate `withReset`. `fallbackFontSize` is named poorly (suggestion: `defaultSliderValue`) but probably needs to be there.

noisysocks committed:
Reuse types from FontSizePickerProps where possible

noisysocks committed:
Rename 'number' to 'parsedValue'

noisysocks committed:
Use constants for defaultOption and customOption

noisysocks committed:
Don't show 'Custom' option in select when custom sizes are disabled

noisysocks committed:
Swap label and aria-label

noisysocks committed:
Use slug in select if name is an empty string

noisysocks committed:
Add test that checks all the t-shirt labels

noisysocks commented:
Feedback addressed or responded to! Thanks again.

ciampo commented:
Thank you! I will try to have another look at this before EOW

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -244,6 +244,27 @@ describe( 'FontSizePicker', () => {
+				for ( const label of [ 'S', 'M', 'L', 'XL', 'XXL' ] ) {
+					const element = screen.getByLabelText( label );
+					expect( element ).toBeInTheDocument();
+					expect( element.children[ 0 ].textContent ).toBe( label );
```
Let's use the `toHaveTextContent` selector instead. Also, no need to drill down the DOM

```suggestion
					expect( element ).toHaveTextContent( label );
```

ciampo reviewed (commented):
There's a couple of comments still in need of addressing, but in the meantime I gave the component a look on Storybook and it looks fine!

I still think that we could add more tests to `FontSizePicker`, given its complexity and the number of edge cases (an example of the need for more coverage is that my manual code review flagged at least 3 instances in which the original logic was not refactored correctly). It would be a good investment , since from what I understand this component will be changed/improved further in the upcoming weeks.

(ps: for the same reasons, more manual testing is welcome too!)

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into update/simplify-font-size-picker

noisysocks committed:
Make filenames match component names

noisysocks committed:
Make DEFAULT_OPTION logic more explicit

noisysocks committed:
Restore headerHint logic to how it was

noisysocks committed:
Use expect().toHaveContent

noisysocks commented:
I rebased this and addressed all the inline feedback. I'll improve the `FontSizePicker` tests tomorrow in a seperate PR so that we can merge that one first.

noisysocks committed:
Rewrite FontSizePicker unit tests to use userEvent and be more comprehensive

noisysocks committed:
Use header label's aria-label

noisysocks committed:
Put back the Gigantosaurus 🦖

noisysocks committed:
Update changelog

noisysocks committed:
Merge branch 'trunk' into update/font-size-picker-unit-tests

noisysocks committed:
Remove editorialising comment

noisysocks committed:
Use getByLabelText( 'Custom Size' ) to select slider

noisysocks committed:
Remove redundant test

noisysocks committed:
Add tests for non-px font sizes

noisysocks committed:
Assert length of options/radios

noisysocks committed:
Assert number of calls to onChange

noisysocks committed:
Merge branch 'update/font-size-picker-unit-tests' into update/simplify-font-size-picker

noisysocks committed:
Can un-comment this failing assertion now

noisysocks committed:
Update CHANGELOG.md

noisysocks commented:
I stacked this PR onto https://github.com/WordPress/gutenberg/pull/45298/.

noisysocks committed:
Merge branch 'trunk' into update/simplify-font-size-picker

noisysocks commented:
Cool, the unit tests added in https://github.com/WordPress/gutenberg/pull/45298 pass! That should reassure us that this refactor doesn't break `FontSizePicker`. How are you feeling @ciampo?

ciampo reviewed (commented):
> How are you feeling @ciampo?

Feeling pretty good about this change and the newly added test — they definitely give us more confidence when making changes moving forward!

Although I still found some discrepancies between this PR and `trunk`, which I outlined in one of the inline conversations. I believe that once that's solved, we'll be ready to merge!

noisysocks committed:
Fix regression when value=""

noisysocks commented:
Nice catch @ciampo, I've fixed that and added regression tests.

ramonjd commented:
> He only needed to change one line but the design of this component tricked him into thinking it was more complicated than it was. 

I tricked myself, I think :)

ramonjd commented on packages/components/src/font-size-picker/utils.ts:
```diff
@@ -56,77 +21,3 @@ export function isSimpleCssValue(
-	} );
-}
-
-export function getSelectedOption(
```
Heads up, it's possible that I'll need this again for https://github.com/WordPress/gutenberg/pull/44967

Maybe not. Either way, it's going to be a gnarly rebase if this goes in first.

🐎

ramonjd replied:
Oh wait, I can probably use `selectedFontSize` or something similar. Never mind me.

noisysocks replied:
Yeah `selectedFontSize` should work. Or you can `fontSizes.find()`. Happy to help you with the rebase if it's too gnarly.

ciampo reviewed (approved):
LGTM 🚀 

Great, great work here. Thank you so much!
